src/coreclr/jit/objectalloc.cpp

```diff
@@ -1335,3 +1483,1734 @@ void ObjectAllocator::RewriteUses()
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(*:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!addr->OperIs(GT_LCL_VAR))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), ...))\n");
+        return nullptr;
+    }
+
+    if (!op2->IsIconHandle(GTF_ICON_CLASS_HDL))
+    {
+        JITDUMP("... no JTRUE(cmp(ind:int(lcl:ref), clsHnd))\n");
+        return nullptr;
+    }
+
+    // Passed the checks... fill in the info.
+    //
+    info->m_local  = addr->AsLclVar()->GetLclNum();
+    bool isNonNull = false;
+    bool isExact   = false;
+    info->m_type   = (CORINFO_CLASS_HANDLE)op2->AsIntCon()->gtIconVal;
+
+    JITDUMP("... " FMT_BB " is guard for V%02u\n", block->bbNum, info->m_local);
+    return tree;
+}
+
+//------------------------------------------------------------------------------
+// CheckForGuardedUse - see if this use of lclNum is controlled by a failing
+//    GDV check that we're tracking as part of conditional escape.
+//
+// Arguments:
+//    block  - block containing tree
+//    tree   - parent tree using the local
+//    lclNum - local being read
+//
+// Returns:
+//    true if this use is a conditionally escaping use.
+//
+bool ObjectAllocator::CheckForGuardedUse(BasicBlock* block, GenTree* tree, unsigned lclNum)
+{
+    bool isGuardedUse = false;
+
+    // Find pseudo local...
+    //
+    unsigned pseudoLocal = BAD_VAR_NUM;
+    if (m_EnumeratorLocalToPseudoLocalMap.TryGetValue(lclNum, &pseudoLocal))
+    {
+        // Verify that this call is made under a **failing** GDV test under the same
+        // conditions tracked by pseudoLocal.
+        //
+        GuardInfo info;
+        if (IsGuarded(block, tree, &info, /* testOutcome */ false))
+        {
+            CloneInfo* pseudoGuardInfo;
+            if (m_CloneMap.Lookup(pseudoLocal, &pseudoGuardInfo))
+            {
+                if ((info.m_local == lclNum) && (pseudoGuardInfo->m_local == lclNum) &&
+                    (info.m_type == pseudoGuardInfo->m_type))
+                {
+                    // If so, track this as an assignment pseudoLocal = ...
+                    //
+                    // Later if we don't clone and split off the failing GDV paths,
+                    // we will mark pseudoLocal as escaped, and that will lead
+                    // to lclNum escaping as well.
+                    //
+                    JITDUMP("... under GDV; tracking via pseudo-local P%02u\n", pseudoLocal);
+                    AddConnGraphEdge(pseudoLocal, lclNum);
+                    isGuardedUse = true;
+                }
+                else
+                {
+                    JITDUMP("... under different guard?\n");
+                }
+            }
+            else
+            {
+                JITDUMP("... under non-gdv guard?\n");
+            }
+        }
+        else
+        {
+            JITDUMP("... not guarded?\n");
+        }
+    }
+    else
+    {
+        JITDUMP("... no pseudo local?\n");
```

Comment:
Seems this would be more readable with some early outs.

Reply:
Revised